sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamAggregationRel.java

```diff
@@ -243,55 +247,72 @@ private Transform(
+            .apply(combiner)
+            .apply(
+                "mergeRecord",
+                ParDo.of(
+                    mergeRecord(outputSchema, windowFieldIndex, ignoreValues, verifyRowValues)))
+            .setRowSchema(outputSchema);
+      }
+      org.apache.beam.sdk.schemas.transforms.Group.AggregateCombinerInterface<Row> globally =
+          org.apache.beam.sdk.schemas.transforms.Group.globally();
+      PTransform<PCollection<Row>, PCollection<Row>> combiner = createCombiner(globally);
+      return windowedStream.apply(combiner).setRowSchema(outputSchema);
+    }
+
+    private PTransform<PCollection<Row>, PCollection<Row>> createCombiner(
+        org.apache.beam.sdk.schemas.transforms.Group.AggregateCombinerInterface<Row>
+            initialCombiner) {
 
-      org.apache.beam.sdk.schemas.transforms.Group.ByFields<Row> byFields =
-          org.apache.beam.sdk.schemas.transforms.Group.byFieldIds(keyFieldsIds);
-      org.apache.beam.sdk.schemas.transforms.Group.CombineFieldsByFields<Row> combined = null;
+      org.apache.beam.sdk.schemas.transforms.Group.AggregateCombinerInterface<Row> combined = null;
       for (FieldAggregation fieldAggregation : fieldAggregations) {
         List<Integer> inputs = fieldAggregation.inputs;
         CombineFn combineFn = fieldAggregation.combineFn;
-        if (inputs.size() > 1 || inputs.isEmpty()) {
-          // In this path we extract a Row (an empty row if inputs.isEmpty).
+        if (inputs.size() == 1) {
+          // Combining over a single field, so extract just that field.
           combined =
               (combined == null)
-                  ? byFields.aggregateFieldsById(inputs, combineFn, fieldAggregation.outputField)
-                  : combined.aggregateFieldsById(inputs, combineFn, fieldAggregation.outputField);
+                  ? initialCombiner.aggregateField(
+                      inputs.get(0), combineFn, fieldAggregation.outputField)
+                  : combined.aggregateField(inputs.get(0), combineFn, fieldAggregation.outputField);
         } else {
-          // Combining over a single field, so extract just that field.
+          // In this path we extract a Row (an empty row if inputs.isEmpty).
           combined =
               (combined == null)
-                  ? byFields.aggregateField(inputs.get(0), combineFn, fieldAggregation.outputField)
-                  : combined.aggregateField(inputs.get(0), combineFn, fieldAggregation.outputField);
+                  ? initialCombiner.aggregateFieldsById(
+                      inputs, combineFn, fieldAggregation.outputField)
+                  : combined.aggregateFieldsById(inputs, combineFn, fieldAggregation.outputField);
         }
       }
 
-      PTransform<PCollection<Row>, PCollection<Row>> combiner = combined;
-      boolean ignoreValues = false;
+      PTransform<PCollection<Row>, PCollection<Row>> combiner =
+          (PTransform<PCollection<Row>, PCollection<Row>>) combined;
```

Comment:
It looks like the problem is that `CombineFieldsGlobally` and `CombineFieldsByFields` extend `PTransform<PCollection<InputT>, PCollection<Row>>`, but `Global` extends `PTransform<PCollection<InputT>, PCollection<Iterable<InputT>>>`. However, it looks like you only use `Global.globally().aggregateField(...)`, which returns a `CombineFieldsGlobally` transform anyway. https://github.com/apache/beam/blob/3a7b8e757d0c220b5838ca6a0805ee463b24f85d/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Group.java#L171

So can you try using `CombineFieldsGlobally` as the initial combiner instead of `Global`? Then `Global` doesn't have to extend `AggregateCombiner` at all.

Reply:
I did this, but I had to create a new method to make an instance of `CombineFieldsGlobally` with null `schemaAggregateFn` because that param is initialized here https://github.com/apache/beam/blob/22e2adbc05cdb3cb2f9800508108926df67f829b/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamAggregationRel.java#L284

Reply:
If we always turn the null `schemaAggregateFn` into `SchemaAggregateFn.create()`, why not initialize with `SchemaAggregateFn.create()` instead of null?

Reply:
I made the change and it worked, thanks!